### ISSUE TITLE
[REVIEW]: Hypothesis: A new approach to property-based testing

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1891]
**Submitting author:** @DRMacIver (<a href="http://orcid.org/0000-0002-8635-3223">David MacIver</a>)
**Repository:** <a href="https://github.com/HypothesisWorks/hypothesis/" target ="_blank">https://github.com/HypothesisWorks/hypothesis/</a>
**Version:** 4.45.0
**Editor:** @danielskatz
**Reviewer:** @luizirber, @djmitche
**Archive:** <a href="https://doi.org/10.5281/zenodo.3548997" target="_blank">10.5281/zenodo.3548997</a>

## Status

[![status](https://joss.theoj.org/papers/e429a3d476392c43bf6b35fe1ceffda1/status.svg)](https://joss.theoj.org/papers/e429a3d476392c43bf6b35fe1ceffda1)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/e429a3d476392c43bf6b35fe1ceffda1"><img src="https://joss.theoj.org/papers/e429a3d476392c43bf6b35fe1ceffda1/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/e429a3d476392c43bf6b35fe1ceffda1/status.svg)](https://joss.theoj.org/papers/e429a3d476392c43bf6b35fe1ceffda1)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/HypothesisWorks/hypothesis/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@luizirber & @djmitche, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/HypothesisWorks/hypothesis/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@DRMacIver) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/HypothesisWorks/hypothesis/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@DRMacIver) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @luizirber, @djmitche it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01891/joss.01891/10.21105.joss.01891.pdf)

**Comment by danielskatz** [comment]
👋 @luizirber and @djmitche - as I think you know, we'll carry out the review here - please carefully read the comments above, and get started when you can - check off items in your checklist, and open issues in the source repo, referring to this issue, when you see things that need to be changed by the authors.

And again, thanks very much for your quick and positive response to the review request.

**Comment by danielskatz** [comment]
If you have any issues or concerns, please let me know here

**Comment by djmitche** [comment]
@danielskatz I think I see an error in the paper itself ("exception" instead of "extension"), but I'm not sure where to find the source for that paper.  Pointers?

**Comment by danielskatz** [comment]
see paper.md and paper.bib in the source repo (root level)

**Comment by djmitche** [comment]
Thanks!  Aside from that (minor) issue I don't see any problems here.

**Comment by danielskatz** [comment]
Thanks @djmitche!

**Comment by luizirber** [comment]
Only one nitpick: There is nothing wrong with the LICENSE.txt file per se, but if you put the content of the MPL 2.0 license linked it will show up as `MPL 2.0` in the repository landing page (instead of `View license`, like it is now).

All my checkboxes are checked.

And I must say this was the easiest review I ever did, but it was somewhat expected because I tend to look for hypothesis dev practices as reference for my projects. The paper is also well written and does a great intro to property testing for scientific projects. Kudos @DRMacIver and @Zac-HD!

**Comment by Zac-HD** [comment]
Thanks @luizirber :tada: 

For the licence, I think this is mostly an effect of Github not having fantastic support for monorepos... [`hypothesis-python/LICENSE.txt`](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/LICENSE.txt) is the file that gets packaged up, and it's the standard MPL2.0 text.  We also use the `setuptools` license classifier, and have a copyright+license header in every `.py` file :smile:

**Comment by Zac-HD** [comment]
@danielskatz - I think we're done, all boxes checked and we've merged the suggested wording change 😁

**Comment by danielskatz** [comment]
Thanks all - I'll continue the process, first checking the license issue, then proof-reading and addressing the final steps

**Comment by danielskatz** [comment]
👋 @arfon - Can you take a look at the license discussion here and the files in the repo and confirm that this is ok, or if not, suggest what changes need to be made?

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01891/joss.01891/10.21105.joss.01891.pdf)

**Comment by danielskatz** [comment]
👋 @arfon - Also, Is there any issue with the author list?

**Comment by arfon** [comment]
> 👋 @arfon - Also, is there any issue with the author list?

It's unconventional, but I think it's probably OK.

> 👋 @arfon - Can you take a look at the license discussion here and the files in the repo and confirm that this is ok, or if not, suggest what changes need to be made?

I also think this is OK.

**Comment by danielskatz** [comment]
@DRMacIver & @Zac-HD - regarding the wording in the paper:

- "lifts a major difficulty" is an odd phrase to me - could we uses resolves or addresses instead of lifts?

Also, please see https://github.com/HypothesisWorks/hypothesis/pull/2212

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1145/351240.351266 is OK
- 10.1145/1159789.1159792 is OK
- 10.1109/32.988498 is OK
- 10.1145/2254064.2254104 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1145/3092703.3092711 is OK
- 10.1145/3092703 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01891/joss.01891/10.21105.joss.01891.pdf)

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01891/joss.01891/10.21105.joss.01891.pdf)

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01891/joss.01891/10.21105.joss.01891.pdf)

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
No archive DOI set. Exiting...

**Comment by danielskatz** [comment]
👋  @DRMacIver & @Zac-HD  - at this point, can you archive the software somewhere (e.g., Zenodo) and let me know the DOI.  Please be sure the metadata (title and authors) associated with the archive match the JOSS paper.

Also, if the version number has changed, please let me know the updated version associated with the archived software.

**Comment by Zac-HD** [comment]
We've been archiving at https://doi.org/10.5281/zenodo.1412597 for a while now!  That's via the github integration though, so it automatically picks up all contributors (same two main authors though). 

The version number has been bumped, but we're actually mid-automatic-release... give it half an hour and we'll be at 4.45.0 😁

**Comment by DRMacIver** [comment]
> Also, if the version number has changed, please let me know the updated version associated with the archived software.

Ah this is something I meant to mention. Hypothesis versioning is very "If you don't like this version number, wait half an hour and there will be a new one" because of our automated release system. We've got a strong backwards compatibility policy and not much in this paper depends on the specific version, but I wonder if there's some way we could indicate that the paper is true of a version range? e.g. `>= 4.45.0, < 5.0.0`. It doesn't matter much and we can pick whichever version is out at point of paper release though.

**Comment by danielskatz** [comment]
We need the DOI of a single version of the code, and the corresponding version number.

I suggest we use http://doi.org/10.5281/zenodo.3548997 and version 4.45.0 

Ok?

**Comment by Zac-HD** [comment]
That's as good as anything else 😁

**Comment by danielskatz** [comment]
@whedon set 4.45.0 as version

**Comment by whedon** [comment]
OK. 4.45.0 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.3548997 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3548997" target="_blank">10.5281/zenodo.3548997</a> is the archive.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1145/351240.351266 is OK
- 10.1145/1159789.1159792 is OK
- 10.1109/32.988498 is OK
- 10.1145/2254064.2254104 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1145/3092703.3092711 is OK
- 10.1145/3092703 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1118

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1118, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
👋 @arfon - in the generated xml file, I see 
```<unstructured_citation>In Praise of Property-Bsed Testing, ://increment.com/testing/in-praise-of-property-based-testing/, 2019, MacIver, David R.</unstructured_citation>```  I realize there's an ```a``` missing in the title, but the URL starting with ```://``` also seems odd - is this what is supposed to be in the XML?  (the URL in the pdf is fine)

**Comment by arfon** [comment]
@danielskatz - this looks like a bug. Let's use Whedon to deposit metadata with Crossref here and I'll file a bug to investigate later.

**Comment by danielskatz** [comment]
ok, will proceed once the missing a PR is merged

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01891/joss.01891/10.21105.joss.01891.pdf)

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1145/351240.351266 is OK
- 10.1145/1159789.1159792 is OK
- 10.1109/32.988498 is OK
- 10.1145/2254064.2254104 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1145/3092703.3092711 is OK
- 10.1145/3092703 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1119

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1119, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by danielskatz** [comment]
👋 @luizirber & @djmitche - thanks for doing your reviews quite quickly, and sorry for all the extra later traffic on this thread

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1197550328180285443) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1120
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01891](https://doi.org/10.21105/joss.01891)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
👋 @DRMacIver & @Zac-HD - thanks for being quite responsive, and congratulations on your publication!

(We'll close this issue shortly)

**Comment by danielskatz** [comment]
👋 @arfon - can you please check/fix the crossref xml data related to the url of the citation we discussed above?

**Comment by danielskatz** [comment]
👋 @arfon - I'm going to close this now, but feel free to reopen it if you need to re the crossref issue

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01891/status.svg)](https://doi.org/10.21105/joss.01891)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01891">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01891/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01891/status.svg
   :target: https://doi.org/10.21105/joss.01891
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01891/status.svg)](https://doi.org/10.21105/joss.01891)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by djmitche** [comment]
This is but a tiny sliver of the email I get from GitHub every day :)

Thanks to all for your hard work on JOSS :)
